bayesian/utility/v2/ca_response_builder.py

```diff
@@ -159,26 +159,25 @@ def get_version_without_cves(self, latest_non_cve_versions):
 
         :return: Highest Version out of all latest_non_cve_versions.
         """
-        logger.info("All Versions with Vulnerabilities.")
-        input_version_tuple = version_info_tuple(
-            convert_version_to_proper_semantic(self.version)
-        )
         highest_version = ''
-        for version in latest_non_cve_versions:
-
-            graph_version_tuple = version_info_tuple(
-                convert_version_to_proper_semantic(version)
-            )
-            if graph_version_tuple > input_version_tuple:
-                if not highest_version:
-                    highest_version = version
-                highest_version_tuple = version_info_tuple(
-                    convert_version_to_proper_semantic(highest_version)
-                )
-                # If version to recommend is closer to what a user is using then, use less than
-                # If recommendation is to show highest version then, use greater than
-                if graph_version_tuple > highest_version_tuple:
-                    highest_version = version
+        try:
+            input_version_comparable = ComparableVersion(self.version)
+            for version in latest_non_cve_versions:
```

Comment:
latest non cve version will never return a list/set. Its always a single value. Why are we using a loop here?

Reply:
In the existing flow this is function variable and comes as a list. Below was the code that populates this parameter:
`latest_non_cve_versions = result_data[0].get('package', {}).get('latest_non_cve_version', [])`

Reply:
latest_non_cve_version = mgmt.makePropertyKey('latest_non_cve_version').dataType(String.class).make();
 schema for this value. it will not allow to store a list

Reply:
We use gremlin to fetch the data, below is batch gremlin query we are using in CA POST flow.

ca_batch_query = """
            epv = [];packages.each {g.V().has('pecosystem', ecosystem).has('pname', it.name)
            .has('version', it.version).as('version', 'cve').select('version').in('has_version')
            .dedup().as('package').select('package', 'version', 'cve')
            .by(valueMap()).by(valueMap()).by(out('has_snyk_cve')
            .valueMap().fold()).fill(epv);};epv;
            """

Out of this query provides packages with non-cve version as an list, though it has always one value, but technically its a list.

Reply:
In fact most of the properties are list as shown in snypet of response below:

"package": {
                   "ecosystem": [
                        "npm"
                    ],
                    "gh_subscribers_count": [
                        5
                    ],
                    "gh_contributors_count": [
                        -1
                    ],
                    "latest_version_last_updated": [
                        "20191117"
                    ],
                    "vertex_label": [
                        "Package"
                    ],
                    "libio_dependents_repos": [
                        "634788"
                    ],
                    "latest_non_cve_version": [
                        "7.0.3"
                    ],
                    "gh_issues_last_year_opened": [
                        -1
                    ],
                    "gh_issues_last_month_closed": [
                        -1
                    ],
                    "gh_open_issues_count": [
                        5
                    ],
                    "libio_dependents_projects": [
                        "290"
                    ],
                    "latest_version": [
                        "7.0.3"
                    ],
                    "tokens": [
                        "cliui"
                    ],
                    "package_relative_used": [
                        "not used"
                    ],
                    "gh_stargazers": [
                        210
                    ],
                    "gh_forks": [
                        16
                    ],
                    "package_dependents_count": [
                        -1
                    ],
                    "gh_prs_last_month_opened": [
                        -1
                    ],
                    "gh_issues_last_year_closed": [
                        -1
                    ],
                    "last_updated": [
                        1.6046478431252089E9
                    ],

Reply:
in that case you can pick [0]

Reply:
@yzainee I would have not touched the existing logical flow, however as per above discussion I have removed latest non-cve into simple comparison check. PTAL